### PR TITLE
Limit users ability to upgrade clusters

### PR DESCRIPTION
**Pull request by anvddriesch**
Towards https://github.com/giantswarm/giantswarm/issues/15345

### CODE REVIEW

**Review by calvix — APPROVED** [review]

File: pkg/aws/cluster/validate_cluster_test.go [hunk 1]
```diff
@@ -33,23 +33,39 @@ func TestValidateReleaseVersion(t *testing.T) {
...
+			valid:             false,
+		},
+		{
+			// version changed with major skip
+			name: "case 4",
+			ctx:  context.Background(),
+
+			newReleaseVersion: "102.0.0",
+			valid:             false,
+		},
```
Comment (xh3b4sd) [inline pkg/aws/cluster/validate_cluster_test.go:62]: With this we see that the test is not very informative. Such a table driven test would better inform about input and output if all inputs would be covered directly. I am wondering what the old release version is now.

**Review by xh3b4sd — APPROVED** [review]
That should work. Just left a comment on test design.

File: pkg/aws/cluster/validate_cluster_test.go (hunk 1 above)
Comment (anvddriesch) [reply pkg/aws/cluster/validate_cluster_test.go:62]: Since the old release version is the same for all tests, I did not include it. But if you feel like it adds a value I can include it.

Comment (xh3b4sd) [reply pkg/aws/cluster/validate_cluster_test.go:62]: For you who wrote the tests that is fine. Though you work in a distributed team and nobody looking at the tests now knows which version is worked with in comparison. I just wanted to point out that in general you want to be as clear as possible in these cases so that you see at a glance what is input and what is output. Then a test is easy to understand and maintain.

Comment (anvddriesch) [reply pkg/aws/cluster/validate_cluster_test.go]: I changed it from using the default version to choosing the old release version explicitly.

File: pkg/aws/cluster/validate_cluster_test.go [hunk 2]
```diff
@@ -25,31 +26,53 @@ func TestValidateReleaseVersion(t *testing.T) {
...
+			newReleaseVersion: "3.3.0",
+			valid:             false,
+		},
+		{
+			// version changed with major skip
+			name: "case 4",
+			ctx:  context.Background(),
+
+			oldReleaseVersion: "3.0.0",
+			newReleaseVersion: "5.0.0",
```
Comment (xh3b4sd) [inline pkg/aws/cluster/validate_cluster_test.go:66]: And now we would want to verify that skipping two or more is also covered. 🙂

Comment (anvddriesch) [reply pkg/aws/cluster/validate_cluster_test.go:66]: There are countless additional cases we could cover here. (what if we skip 2 releases? 3? 5?) However, I think it's good to keep the test cases basic and representative. If we find edge cases that we want to cover in the future, let"s add them here. However, this was not such a case since the upgrade validation is based on a simple `a greater-than b+1` condition that should remain true regardless of how many majors we skip.

Comment (xh3b4sd) [reply pkg/aws/cluster/validate_cluster_test.go:66]: We do not have to cover everything, that is true. I for myself I just try to cover the obvious cases and since I saw some of them missing now after the tests became readable I suggested more cases. If you are fine with your test cases, you take the responsibility of ensuring this component does what it should even in two years down the road. I am ok with that. 🙂
